docs/man/xrdp.ini.5.in
.TH "xrdp.ini" "5" "@PACKAGE_VERSION@" "xrdp team" ""
.SH "NAME"
\fBxrdp.ini\fR \- Configuration file for \fBxrdp\fR(8)

.SH "DESCRIPTION"
This is the man page for \fBxrdp.ini\fR, \fBxrdp\fR(8) configuration file.
It is composed by a number of sections, each one composed by a section name, enclosed by square brackets, followed by a list of \fI<parameter>\fR=\fI<value>\fR lines.

\fBxrdp.ini\fR supports the following sections:

.TP
\fB[Globals]\fP \- sets some global configuration settings for \fBxrdp\fR(8).

.TP
\fB[Logging]\fP \- logging subsystem parameters

.TP
\fB[Channels]\fP \- channel subsystem parameters

.LP
All options and values (except for file names and paths) are case insensitive, and are described in detail below.

.SH "GLOBALS"
The options to be specified in the \fB[Globals]\fR section are the following:

.TP
\fBport\fR=\fIport_specification [ port_specification ... ]\fR
Specify the port(s) that xrdp should listen on. More instructions and
examples can be found within xrdp.ini itself.

.TP
\fBinstance_name\fR=\fI<arbitrary-string>\fR
Specify an arbitrary name for the xrdp instance that can be used to
identify sessions that this xrdp instance started, or which were
spun up with the same name using \fBxrdp-sesrun\fR. When using the
\fBN\fR policy in \fBsesman.ini\fR these sessions are kept distinct
and can be reconnected to by connecting to the respective \fBxrdp\fR
instance.

.TP
\fBvmconnect\fR=\fI[true|false]\fR
(Hyper-V VMs only). Enables a wider support of security protocols when a
virtual machine running xrdp is hosted on Hyper-V, and the user connects
to it via the vmms service on TCP port 2179. In this configuration,
RDP security is handled by the vmms service, and security features which
are not yet added to xrdp itself can be supported.

This parameter is required in environments which do not support 'classic'
RDP encryption.

The parameter is ignored for connections which are not made to the
virtual machine over the vsock transport.

.TP
\fBautorun\fP=\fIsession_name\fP
Section name for automatic login. If set and the client supplies valid
username and password, the user will be logged in automatically using the
connection specified by \fIsession_name\fP.

If \fIsession_name\fP is empty, the \fBLOGIN DOMAIN\fR from the client
with be used to select the section. If no domain name is supplied, the
first suitable section will be used for automatic login.

.TP
\fBbitmap_cache\fR=\fI[true|false]\fR
If set to \fB1\fR, \fBtrue\fR or \fByes\fR this option enables bitmap caching in \fBxrdp\fR(8).

.TP
\fBbitmap_compression\fR=\fI[true|false]\fR
If set to \fB1\fR, \fBtrue\fR or \fByes\fR this option enables bitmap compression in \fBxrdp\fR(8).

.TP
\fBbulk_compression\fP=\fI[true|false]\fP
If set to \fB1\fR, \fBtrue\fR or \fByes\fR this option enables compression of bulk data in \fBxrdp\fR(8).

.TP
\fBcertificate\fP=\fI/path/to/certificate\fP
.TP
\fBkey_file\fP=\fI/path/to/private_key\fP
Set location of TLS certificate and private key. They must be written in PEM format.
If not specified, defaults to \fB@sysconfdir@/@sysconfsubdir@/cert.pem\fP, \fB@sysconfdir@/@sysconfsubdir@/key.pem\fP.

This parameter is effective only if \fBsecurity_layer\fP is set to \fBtls\fP or \fBnegotiate\fP.

.TP
\fBtls_pms_log_file\fR=\fI<path-to-log-file>\fR
Logs TLS pre-master secrets to the specified file. This allows packet capture
tools (e.g. Wireshark) to decrypt captured PDUs.

The file must be writeable by xrdp and readable by the packet capture tool.
A good way to achive this is to create a temporary directory with
permissions 2750 owned by the user running xrdp, and in the group used by
the packet sniffer.

SETTING THIS OPTION IS A SECURITY RISK. ONLY SET THIS OPTION FOR DEBUGGING
COMMUNICATIONS BETWEEN XRDP AND A CLIENT.

.TP
\fBchannel_code\fP=\fI[true|false]\fP
If set to \fB0\fR, \fBfalse\fR or \fBno\fR this option disables all channels \fBxrdp\fR(8).
See section \fBCHANNELS\fP below for more fine grained options.

.TP
\fBcrypt_level\fP=\fI[low|medium|high|fips]\fP
.\" <http://blogs.msdn.com/b/openspecification/archive/2011/12/08/encryption-negotiation-in-rdp-connection.aspx>
Regulate encryption level of Classic RDP Security.
This parameter is effective only if \fBsecurity_layer\fP is set to \fBrdp\fP or \fBnegotiate\fP.

Encryption in Classic RDP Security is controlled by two settings: \fIEncryption Level\fP
and \fIEncryption Method\fP.  The only supported \fIEncryption Method\fP are \fB40BIT_ENCRYPTION\fP
and \fB128BIT_ENCRYPTION\fP. \fB56BIT_ENCRYPTION\fP is not supported.
This option controls the \fIEncryption Level\fP:
.RS 8
.TP
.B low
All data sent from the client to the server is protected by encryption based on
the maximum key strength supported by the client.
.I This is the only level that the traffic sent by the server to client is not encrypted.
.TP
.B medium
All data sent between the client and the server is protected by encryption based on
the maximum key strength supported by the client (client compatible).
.TP
.B high
All data sent between the client and the server is protected by encryption based on
the server's maximum key strength (sever compatible).
.TP
.B fips
All data sent between the client and server is protected using Federal Information
Processing Standard 140-1 validated encryption methods. Note that FIPS 140-1 is
no longer considered secure.
.I This level is required for Windows clients (mstsc.exe) if the client's group policy
.I enforces FIPS-compliance mode.
.RE

.TP
\fBfork\fP=\fI[true|false]\fP
If set to \fB1\fR, \fBtrue\fR or \fByes\fR for each incoming connection \fBxrdp\fR(8) forks a sub-process instead of using threads.

.TP
\fBhidelogwindow\fP=\fI[true|false]\fP
If set to \fB1\fP, \fBtrue\fP or \fByes\fP, \fBxrdp\fP will not show a window for log messages.
If not specified, defaults to \fBfalse\fP.

.TP
\fBmax_bpp\fP=\fI[8|15|16|24|32]\fP
Limit the color depth by specifying the maximum number of bits per pixel.
If not specified or set to \fB0\fP, unlimited.

.TP
\fBpamerrortxt\fP=\fIerror_text\fP
Specify additional text displayed to user if authentication fails. The maximum length is \fB256\fP.

The use of 'pam' in the name of this option is historic

.TP
\fBport\fP=\fIport\fP
Specify TCP port and interface to listen on for incoming connections.
Specifying only the port means that xrdp will listen on all interfaces.
The default port for RDP is \fB3389\fP.
Multiple address:port instances must be separated by spaces or commas. Check the .ini file for examples.
Specifying interfaces requires said interfaces to be UP before xrdp starts.

.TP
\fBruntime_user\fP=\fIusername\fP
.TP
\fBruntime_group\fP=\fIgroupname\fP
User name and group to run the xrdp daemon under.

After xrdp starts, it sets its UID and GID to values derived from these
settings, so that it's running without system privilege.

The \fBruntime_group\fP MUST be set to the same value as
\fBSessionSockdirGroup\fP in \fBsesman.ini\fP if you want to run sessions.

A suitable user and group can be added with a command like this (Linux):-

useradd xrdp -d / -c 'xrdp daemon' -s /usr/sbin/nologin

In order to establish secure connections, the xrdp daemon needs permission
to access sensitive cryptographic files. After changing either or both
of these values, check that xrdp has access to required files by running
this script:-

@xrdpdatadir@/xrdp-chkpriv

.TP
\fBenable_token_login\fP=\fI[true|false]\fP
If set to \fB1\fP, \fBtrue\fP or \fByes\fP, \fBxrdp\fP will scan the user name provided by the
client for the ASCII field separator character (0x1F). It will then copy over what is after the
separator as the password supplied by the user and treats it as autologon. If not specified,
defaults to \fBfalse\fP.

.TP
\fBdomain_user_separator\fP=\fBseparator\fP
If specified the domain name supplied by the client is appended to the username separated
by \fBseparator\fP.

.TP
\fBrequire_credentials\fP=\fI[true|false]\fP
If set to \fB1\fP, \fBtrue\fP or \fByes\fP, \fBxrdp\fP requires clients
to include username and password initial connection phase. In other
words, xrdp doesn't allow clients to show login screen if set to true.
It follows that an incorrect password will cause the login to immediately
fail without displaying the login screen.  If not specified, defaults
to \fBfalse\fP.

.TP
\fBsecurity_layer\fP=\fI[tls|rdp|negotiate]\fP
Regulate security methods. If not specified, defaults to \fBnegotiate\fP.
.RS 8
.TP
.B tls
Enhanced RDP Security is used. All security operations (encryption, decryption, data integrity
verification, and server authentication) are implemented by TLS.

.TP
.B rdp
Classic RDP Security is used. The encryption level
of Classic RDP Security is controlled by \fBcrypt_level\fP.
Use this setting for testing only.

.TP
.B negotiate
Negotiate these security methods with clients.
.RE

.TP
\fBssl_protocols\fP=\fI[SSLv3] [TLSv1] [TLSv1.1] [TLSv1.2] [TLSv1.3]\fP
Enables the specified SSL/TLS protocols. Each value should be separated by comma.
SSLv2 is always disabled. At least one protocol should be given to accept TLS connections.
This parameter is effective only if \fBsecurity_layer\fP is set to \fBtls\fP or \fBnegotiate\fP.

.TP
\fBtcp_keepalive\fP=\fI[true|false]\fP
Regulate if the listening socket uses socket option \fBSO_KEEPALIVE\fP.
If set to \fB1\fP, \fBtrue\fP or \fByes\fP and the network connection disappears
without closing messages, the connection will be closed.

.TP
\fBtcp_nodelay\fP=\fI[true|false]\fP
Regulate if the listening socket uses socket option \fBTCP_NODELAY\fP.
If set to \fB1\fP, \fBtrue\fP or \fByes\fP, no buffering will be performed in the TCP stack.

.TP
\fBtcp_send_buffer_bytes\fP=\fIbuffer_size\fP
.TP
\fBtcp_recv_buffer_bytes\fP=\fIbuffer_size\fP
Specify send/recv buffer sizes in bytes.  The default value depends on
the operating system. It is recommended not to set these on systems with
dynamic TCP buffer sizing

.TP
\fBtls_ciphers\fP=\fIcipher_suite\fP
Specifies TLS cipher suite. The format of this parameter is equivalent
to which \fBopenssl\fP(1) ciphers subcommand accepts.

(ex. $ openssl ciphers 'HIGH:!ADH:!SHA1')

This parameter is effective only if \fBsecurity_layer\fP is set to \fBtls\fP or \fBnegotiate\fP.

.TP
\fBuse_fastpath\fP=\fI[input|output|both|none]\fP
If not specified, defaults to \fBnone\fP.

.TP
\fBblack\fP=\fI000000\fP
.TP
\fBgrey\fP=\fIc0c0c0\fP
.TP
\fBdark_grey\fP=\fI808080\fP
.TP
\fBblue\fP=\fI0000ff\fP
.TP
\fBdark_blue\fP=\fI00007f\fP
.TP
\fBwhite\fP=\fIffffff\fP
.TP
\fBred\fP=\fIff0000\fP
.TP
\fBgreen\fP=\fI00ff00\fP
.TP
\fBbackground\fP=\fI000000\fP
These options override the colors used internally by \fBxrdp\fP(8) to draw the login and log windows.
Colors are defined using a hexadecimal (hex) notation for the combination of Red, Green, and Blue color values (RGB).
The lowest value that can be given to one of the light sources is 0 (hex 00).
The highest value is 255 (hex FF).
.TP
\fBfv1_select\fP=\fI130:sans-18.fv1,0:sans-10.fv1\fP
Selects a default fv1 font.
This parameter is a comma-separated list of DPI:name pairs.  The list
is scanned from left-to-right. The font used is the first font whose DPI
value is less-than-or-equal to the vertical DPI of the monitor used for
the login screen.
.TP
\fBdefault_dpi\fP=\fI96\fP
Default DPI used for a monitor if the client does not send physical
size information.


.SH "LOGGING"
The following parameters can be used in the \fB[Logging]\fR section:

.TP
\fBLogFile\fR=\fI@localstatedir@/log/xrdp.log\fR
This options contains the path to logfile. It can be either absolute or relative. If set to \fB<stdout>\fR, log will go to stdout. Use for debugging only\fR

.TP
\fBLogLevel\fR=\fIlevel\fR
This option can have one of the following values:

\fBCORE\fR or \fB0\fR \- Log only core messages. these messages are _always_ logged, regardless the logging level selected.

\fBERROR\fR or \fB1\fR \- Log only error messages

\fBWARNING\fR, \fBWARN\fR or \fB2\fR \- Logs warnings and error messages

\fBINFO\fR or \fB3\fR \- Logs errors, warnings and informational messages

\fBDEBUG\fR or \fB4\fR \- Log everything. If \fBxrdp-sesman\fR is compiled in debug mode, this options will output many more low\-level message, useful for developers

.TP
\fBEnableSyslog\fR=\fI[true|false]\fR
If set to \fB1\fR, \fBtrue\fR or \fByes\fR this option enables logging to syslog. Otherwise syslog is disabled.

.TP
\fBSyslogLevel\fR=\fIlevel\fR
This option sets the logging level for syslog. It can have the same values of \fBLogLevel\fR. If \fBSyslogLevel\fR is greater than \fBLogLevel\fR, its value is lowered to that of \fBLogLevel\fR.

.TP
\fBEnableConsole\fR=\fI[true|false]\fR
If set to \fB1\fR, \fBtrue\fR or \fByes\fR, this option enables logging to the console (ie. stdout).

.TP
\fBConsoleLevel\fR=\fIlevel\fR
Logging level for the console. It can have the same values as \fBLogLevel\fR. Defaults to \fBDEBUG\fR.

.TP
\fBEnableProcessId\fR=\fI[true|false]\fR
If set to \fB1\fR, \fBtrue\fR or \fByes\fR, this option enables logging the process id in all log messages. Defaults to \fBfalse\fR.

.SH "CHANNELS"
The Remote Desktop Protocol supports several channels, which are used to transfer additional data like sound, clipboard data and others.
Channel names not listed here will be blocked by \fBxrdp\fP.
Not all channels are supported in all cases, so setting a value to \fItrue\fP is a prerequisite, but does not force its use.
.br

xrdp/xrdp.c
/**
 * xrdp: A Remote Desktop Protocol server.
 *
 * Copyright (C) Jay Sorg 2004-2014
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * main program
 */

#if defined(HAVE_CONFIG_H)
#include <config_ac.h>
#endif

#include <stdarg.h>

#include "xrdp.h"
#include "log.h"
#include "xrdp_configure_options.h"
#include "copying_third_party.h"
#include "string_calls.h"

#if !defined(PACKAGE_VERSION)
#define PACKAGE_VERSION "???"
#endif

static struct xrdp_listen *g_listen = 0;


/*****************************************************************************/
static void
print_license(void)
{
    g_writeln("Third Party Code Additional Copyright Notices and License Terms");
    g_writeln("%s", "");
    g_writeln("Following third-party code are used in xrdp %s:", PACKAGE_VERSION);
    g_writeln("%s", "");
    g_writeln("%s", copying_third_party);
}
/*****************************************************************************/
static void
print_version(void)
{
    g_writeln("xrdp %s", PACKAGE_VERSION);
    g_writeln("  A Remote Desktop Protocol Server.");
    g_writeln("  Copyright (C) 2004-%d Jay Sorg, "
              "Neutrino Labs, and all contributors.",
              VERSION_YEAR);
    g_writeln("  See https://github.com/neutrinolabs/xrdp for more information.");
    g_writeln("%s", "");

#if defined(XRDP_CONFIGURE_OPTIONS)
    g_writeln("  Configure options:");
    g_writeln("%s", XRDP_CONFIGURE_OPTIONS);
#endif

    g_writeln("  Compiled with %s", get_openssl_version());

}

/*****************************************************************************/
static void
print_help(void)
{
    g_writeln("Usage: xrdp [options]");
    g_writeln("   -k, --kill        shut down xrdp");
    g_writeln("   -h, --help        show help");
    g_writeln("   -v, --version     show version");
    g_writeln("   -n, --nodaemon    don't fork into background");
    g_writeln("   -p, --port        tcp listen port");
    g_writeln("   -f, --fork        fork on new connection");
    g_writeln("   -c, --config      specify new path to xrdp.ini");
    g_writeln("       --dump-config display config on stdout on startup");
    g_writeln("       --license     show additional license information");
}

/*****************************************************************************/
/* Signal handler for SIGINT and SIGTERM
 * Note: only signal safe code (eg. setting wait event) should be executed in
 * this function. For more details see `man signal-safety`
 */
static void
xrdp_shutdown(int sig)
{
    g_set_wait_obj(g_get_term());
}

/*****************************************************************************/
/* Signal handler for SIGCHLD
 * Note: only signal safe code (eg. setting wait event) should be executed in
 * this function. For more details see `man signal-safety`
 */
static void
xrdp_child(int sig)
{
    g_set_sigchld(1);
}

/*****************************************************************************/
/**
 * @brief looks for a case-insensitive match of a string in a list
 * @param candidate  String to match
 * @param ... NULL-terminated list of strings to compare the candidate with
 * @return !=0 if the candidate is found in the list
 */
static int nocase_matches(const char *candidate, ...)
{
    va_list vl;
    const char *member;
    int result = 0;

    va_start(vl, candidate);
    while ((member = va_arg(vl, const char *)) != NULL)
    {
        if (g_strcasecmp(candidate, member) == 0)
        {
            result = 1;
            break;
        }
    }

    va_end(vl);
    return result;
}

/*****************************************************************************/
/* No-op signal handler.
 * Note: only signal safe code (eg. setting wait event) should be executed in
 * this function. For more details see `man signal-safety`
 */
static void
xrdp_sig_no_op(int sig)
{
    /* no-op */
}

/*****************************************************************************/
/**
 *
 * @brief  Command line argument parser
 * @param argc number of command line arguments
 * @param argv pointer array of commandline arguments
 * @param startup_params [out] Returned startup parameters
 * @return 0 on success, n on nth argument is unknown
 *
 */
static int
xrdp_process_params(int argc, char **argv,
                    struct xrdp_startup_params *startup_params)
{
    int index;
    const char *option;
    const char *value;

    index = 1;

    while (index < argc)
    {
        option = argv[index];

        if (index + 1 < argc)
        {
            value = argv[index + 1];
        }
        else
        {
            value = "";
        }

        if (nocase_matches(option, "-help", "--help", "-h", NULL))
        {
            startup_params->help = 1;
        }
        else if (nocase_matches(option, "-kill", "--kill", "-k", NULL))
        {
            startup_params->kill = 1;
        }
        else if (nocase_matches(option, "-nodaemon", "--nodaemon", "-n",
                                "-nd", "--nd", "-ns", "--ns", NULL))
        {
            startup_params->no_daemon = 1;
        }
        else if (nocase_matches(option, "-v", "--version", NULL))
        {
            startup_params->version = 1;
        }
        else if (nocase_matches(option, "-p", "--port", NULL))
        {
            index++;
            g_strncpy(startup_params->port, value,
                      sizeof(startup_params->port) - 1);

            if (g_strlen(startup_params->port) < 1)
            {
                g_writeln("error processing params, port [%s]", startup_params->port);
                return 1;
            }
            else
            {
                g_writeln("--port parameter found, ini override [%s]",
                          startup_params->port);
            }
        }
        else if (nocase_matches(option, "-f", "--fork", NULL))
        {
            startup_params->fork = 1;
            g_writeln("--fork parameter found, ini override");
        }
        else if (nocase_matches(option, "--dump-config", NULL))
        {
            startup_params->dump_config = 1;
        }
        else if (nocase_matches(option, "--license", NULL))
        {
            startup_params->license = 1;
        }
        else if (nocase_matches(option, "-c", "--config", NULL))
        {
            index++;
            startup_params->xrdp_ini = value;
        }
        else /* unknown option */
        {
            return index;
        }

        index++;
    }

    return 0;
}

/*****************************************************************************/
/**
 *
 * @brief  Read additional startup parameters from xrdp.ini
 *
 * @param[in,out] startup_params parameters from the command line
 * @return 0 on success
 *
 */
static int
read_xrdp_ini_startup_params(struct xrdp_startup_params *startup_params)
{
    int rv = 0;
    int fd;
    int index;
    int port_override;
    int fork_override;
    const char *name;
    const char *val;
    struct list *names;
    struct list *values;

    port_override = startup_params->port[0] != 0;
    fork_override = startup_params->fork;
    names = list_create();
    names->auto_free = 1;
    values = list_create();
    values->auto_free = 1;

    fd = g_file_open_ro(startup_params->xrdp_ini);
    if (fd < 0)
    {
        LOG(LOG_LEVEL_ERROR, "Can't open %s [%s]", startup_params->xrdp_ini,
            g_get_strerror());
        rv = 1;
    }
    else if (file_read_section(fd, "globals", names, values) != 0)
    {
        LOG(LOG_LEVEL_ERROR, "Can't read [Globals] from %s",
            startup_params->xrdp_ini);
        rv = 1;
    }
    else
    {
        for (index = 0; index < names->count; index++)
        {
            name = (const char *)list_get_item(names, index);
            val = (const char *)list_get_item(values, index);
            if (name == 0 || val == 0)
            {
                continue;
            }

            if (g_strcasecmp(name, "port") == 0)
            {
                if (port_override == 0)
                {
                    g_strncpy(startup_params->port, val,
                              sizeof(startup_params->port) - 1);
                }
            }

            if (g_strcasecmp(name, "instance_name") == 0)
            {
                strlcpy(startup_params->instance_name, val,
                        sizeof(startup_params->instance_name) - 1);
            }

            else if (g_strcasecmp(name, "fork") == 0)
            {
                if (fork_override == 0)
                {
                    startup_params->fork = g_text2bool(val);
                }
            }

            else if (g_strcasecmp(name, "tcp_nodelay") == 0)
            {
                startup_params->tcp_nodelay = g_text2bool(val);
            }

            else if (g_strcasecmp(name, "tcp_keepalive") == 0)
            {
                startup_params->tcp_keepalive = g_text2bool(val);
            }

            else if (g_strcasecmp(name, "tcp_send_buffer_bytes") == 0)
            {
                startup_params->tcp_send_buffer_bytes = g_atoi(val);
            }

            else if (g_strcasecmp(name, "tcp_recv_buffer_bytes") == 0)
            {
                startup_params->tcp_recv_buffer_bytes = g_atoi(val);
            }

            else if (g_strcasecmp(name, "use_vsock") == 0)
            {
                LOG(LOG_LEVEL_WARNING,
                    "Config parameter 'use_vsock' is obsolete. "
                    "Use 'port=vsock://...' instead");
                startup_params->use_vsock = g_text2bool(val);
            }

            else if (g_strcasecmp(name, "runtime_user") == 0)
            {
                g_snprintf(startup_params->runtime_user,
                           sizeof(startup_params->runtime_user),
                           "%s", val);
            }

            else if (g_strcasecmp(name, "runtime_group") == 0)
            {
                g_snprintf(startup_params->runtime_group,
                           sizeof(startup_params->runtime_group),
                           "%s", val);
            }
        }
    }

    list_delete(names);
    list_delete(values);
    if (fd >= 0)
    {
        g_file_close(fd);
    }
    return rv;
}


/*****************************************************************************/
/* Basic sanity checks before any forking */
static int
xrdp_sanity_check(void)
{
    int intval = 1;
    int host_be;
    const char *key_file = XRDP_CFG_PATH "/rsakeys.ini";

    /* check compiled endian with actual endian */
    host_be = !((int)(*(unsigned char *)(&intval)));

#if defined(B_ENDIAN)
    if (!host_be)
    {
        g_writeln("Not a big endian machine, edit arch.h");
        return 1;
    }
#endif
#if defined(L_ENDIAN)
    if (host_be)
    {
        g_writeln("Not a little endian machine, edit arch.h");
        return 1;
    }
#endif

    /* check long, int and void* sizes */
#if SIZEOF_INT != 4
#   error unusable int size, must be 4
#endif

#if SIZEOF_LONG != SIZEOF_VOID_P
#   error sizeof(long) must match sizeof(void*)
#endif

#if SIZEOF_LONG != 4 && SIZEOF_LONG != 8
#   error sizeof(long), must be 4 or 8
#endif

    if (sizeof(tui64) != 8)
    {
        g_writeln("unusable tui64 size, must be 8");
        return 1;
    }

    if (!g_file_exist(key_file))
    {
        g_writeln("File %s is missing, create it using xrdp-keygen", key_file);
        return 1;
    }

    return 0;
}

/*****************************************************************************/
static int
check_drop_privileges(struct xrdp_startup_params *startup_params)
{
    int rv = 1;
    const char *user = startup_params->runtime_user;
    const char *group = startup_params->runtime_group;
    int entry_uid = g_getuid();
    int entry_gid = g_getgid();

    // Regard a user as privileged if either the UID or GID is zero
    int privileged_user = (entry_uid == 0 || entry_gid == 0);
    if (!privileged_user)
    {
        // xrdp daemon UID:GID is already set by whatever started it. We'll
        // log this as it may be unexpected and continue. We don't have
        // any privilege to override this.
        LOG(LOG_LEVEL_INFO,
            "xrdp is already running as an unprivileged user uid=%d gid=%d",
            entry_uid, entry_gid);
        rv = 0;
    }
    else if (user[0] == '\0' && group[0] == '\0')
    {
        // We're running as a privileged user, and this isn't
        // overridden in our own config
        LOG(LOG_LEVEL_ALWAYS,
            "You are running xrdp as uid=%d gid=%d. This is not recommended.",
            entry_uid, entry_gid);
        rv = 0; // Allowed for now for old configs
    }
    else if (user[0] == '\0' || group[0] == '\0')
    {
        LOG(LOG_LEVEL_ERROR,
            "Both a runtime_user and a runtime_group MUST be specified");
    }
    else
    {
        rv = g_drop_privileges(user, group);
        if (rv == 0)
        {
            LOG(LOG_LEVEL_INFO, "Switched user:group to %s:%s", user, group);
        }
    }

    return rv;
}

/*****************************************************************************/
static int
read_pid_file(const char *pid_file)
{
    int pid = -1;
    int fd = g_file_open_ro(pid_file); /* xrdp.pid */
    if (fd >= 0)
    {
        char text[32];
        g_memset(text, 0, sizeof(text));
        g_file_read(fd, text, sizeof(text) - 1);
        pid = g_atoi(text);
        g_file_close(fd);
    }

    return pid;
}

/*****************************************************************************/
/**
 * Kills an active xrdp daemon
 *
 * It is assumed that logging is not active
 *
 * @param pid_file PID file
 * @return 0 for success
 */
static int
kill_daemon(const char *pid_file)
{
    int status = 1;
    int pid;
    if (g_getuid() != 0)
    {
        g_writeln("Must be root");
    }
    else if ((pid = read_pid_file(pid_file)) > 0)
    {
        if (!g_pid_is_active(pid))
        {
            g_writeln("Daemon not active");
            status = 0;
        }
        else
        {
            g_writeln("stopping process id %d", pid);
            int i;
            g_sigterm(pid);
            g_sleep(100);
            i = 5 * 1000 / 500;
            while (i > 0 && g_pid_is_active(pid))
            {
                g_sleep(500);
                --i;
            }

            if (g_pid_is_active(pid))
            {
                g_writeln("Can't stop process");
            }
            else
            {
                status = 0;
            }
        }

        if (status == 0)
        {
            /* delete the xrdp.pid file, as xrdp can't do this
             * if it's running without privilege */
            g_file_delete(pid_file);
        }
    }

    return status;
}

/*****************************************************************************/
int
main(int argc, char **argv)
{
    int exit_status = 1;
    enum logReturns error;
    struct xrdp_startup_params startup_params = {0};
    int pid;
    int daemon_pid = 0;
    int daemon;
    char text[256];
    const char *pid_file = XRDP_PID_PATH "/xrdp.pid";
    int errored_argc;

#ifdef USE_DEVEL_LOGGING
    int test;
    for (test = 0; test < argc; test++)
    {
        g_writeln("Argument %i - %s", test, argv[test]);
    }
#endif

    g_init("xrdp");
    ssl_init();

    startup_params.xrdp_ini = XRDP_CFG_PATH "/xrdp.ini";

    errored_argc = xrdp_process_params(argc, argv, &startup_params);
    if (errored_argc > 0)
    {
        print_version();
        g_writeln("%s", "");
        print_help();
        g_writeln("%s", "");

        g_writeln("Unknown option: %s", argv[errored_argc]);
        g_deinit();
        g_exit(1);
    }

    if (startup_params.help)
    {
        print_version();
        g_writeln("%s", "");
        print_help();

        g_deinit();
        g_exit(0);
    }

    if (startup_params.version)
    {
        print_version();
        g_deinit();
        g_exit(0);
    }

    if (startup_params.license)
    {
        print_license();
        g_deinit();
        g_exit(0);
    }

    if (xrdp_sanity_check() != 0)
    {
        g_writeln("Fatal error occurred, exiting");
        g_deinit();
        g_exit(1);
    }

    if (startup_params.kill)
    {
        int status = kill_daemon(pid_file);
        g_deinit();
        g_exit(status);
    }

    /* starting logging subsystem */
    error = log_start(startup_params.xrdp_ini, "xrdp",
                      (startup_params.dump_config) ? LOG_START_DUMP_CONFIG : 0);

    if (error != LOG_STARTUP_OK)
    {
        switch (error)
        {
            case LOG_ERROR_MALLOC:
                g_writeln("error on malloc. cannot start logging. quitting.");
                break;
            case LOG_ERROR_FILE_OPEN:
                g_writeln("error opening log file [%s]. quitting.",
                          getLogFile(text, 255));
                break;
            case LOG_ERROR_NO_CFG:
                g_writeln("config file %s unreadable or missing",
                          startup_params.xrdp_ini);
                break;
            default:
                g_writeln("log_start error");
                break;
        }

        g_deinit();
        g_exit(1);
    }

    if (read_xrdp_ini_startup_params(&startup_params) != 0)
    {
        log_end();
        g_deinit();
        g_exit(1);
    }

    if ((pid = read_pid_file(pid_file)) > 0 && g_pid_is_active(pid))
    {
        LOG(LOG_LEVEL_ALWAYS,
            "It looks like xrdp (pid=%d) is already running.", pid);
        LOG(LOG_LEVEL_ALWAYS, "If not, delete %s and try again.", pid_file);
        log_end();
        g_deinit();
        g_exit(1);
    }

    daemon = !startup_params.no_daemon;


    if (daemon)
    {
        /* make sure containing directory exists */
        g_create_path(pid_file);

        /* make sure we can write to pid file */
        int pid_fd = g_file_open_rw(pid_file); /* xrdp.pid */

        if (pid_fd == -1)
        {
            LOG(LOG_LEVEL_ALWAYS,
                "running in daemon mode with no access to pid files, quitting");
            log_end();
            g_deinit();
            g_exit(1);
        }

        /* Before daemonising, check we can listen.
         * If we can't listen, exit with failure status */
        struct xrdp_listen *xrdp_listen;
        xrdp_listen = xrdp_listen_create(&startup_params);
        int status = xrdp_listen_init(xrdp_listen);
        xrdp_listen_delete(xrdp_listen);

        if (status != 0)
        {
            LOG(LOG_LEVEL_ALWAYS, "Failed to start xrdp daemon, "
                "possibly address already in use.");
            log_end();
            g_deinit();
            /* must exit with failure status,
               or systemd cannot detect xrdp daemon couldn't start properly */
            g_exit(1);
        }

        /* start of daemonizing code */
        pid = g_fork();

        if (pid == -1)
        {
            LOG(LOG_LEVEL_ALWAYS, "problem forking [%s]", g_get_strerror());
            log_end();
            g_deinit();
            g_exit(1);
        }

        if (0 != pid)
        {
            /* exit, this is the main process */
            log_end();
            g_deinit();
            g_exit(0);
        }

        daemon_pid = g_getpid();
        g_sleep(1000);
        /* write our pid to file */
        g_sprintf(text, "%d", g_getpid());
        g_file_write(pid_fd, text, g_strlen(text));
        g_file_close(pid_fd);

        g_sleep(1000);
        g_file_close(0);
        g_file_close(1);
        g_file_close(2);

        if (g_file_open_rw("/dev/null") < 0)
        {
        }

        if (g_file_open_rw("/dev/null") < 0)
        {
        }

        if (g_file_open_rw("/dev/null") < 0)
        {
        }

        /* end of daemonizing code */
    }

    g_listen = xrdp_listen_create(&startup_params);
    if (xrdp_listen_init(g_listen) != 0)
    {
        LOG(LOG_LEVEL_ALWAYS, "Failed to start xrdp daemon, "
            "possibly address already in use.");
    }
    else if (check_drop_privileges(&startup_params) == 0)
    {
        g_set_threadid(tc_get_threadid());
        g_signal_user_interrupt(xrdp_shutdown); /* SIGINT */
        g_signal_pipe(xrdp_sig_no_op);          /* SIGPIPE */
        g_signal_terminate(xrdp_shutdown);      /* SIGTERM */
        g_signal_child_stop(xrdp_child);        /* SIGCHLD */
        g_signal_hang_up(xrdp_sig_no_op);       /* SIGHUP */
        g_set_sync_mutex(tc_mutex_create());
        g_set_sync1_mutex(tc_mutex_create());
        pid = g_getpid();
        LOG(LOG_LEVEL_INFO, "starting xrdp with pid %d", pid);
        g_snprintf(text, 255, "xrdp_%8.8x_main_term", pid);
        g_set_term_event(g_create_wait_obj(text));

        if (g_get_term() == 0)
        {
            LOG(LOG_LEVEL_WARNING, "error creating g_term_event");
        }

        g_snprintf(text, 255, "xrdp_%8.8x_main_sigchld", pid);
        g_set_sigchld_event(g_create_wait_obj(text));

        if (g_get_sigchld() == 0)
        {
            LOG(LOG_LEVEL_WARNING, "error creating g_sigchld_event");
        }

        g_snprintf(text, 255, "xrdp_%8.8x_main_sync", pid);
        g_set_sync_event(g_create_wait_obj(text));

        if (g_get_sync_event() == 0)
        {
            LOG(LOG_LEVEL_WARNING, "error creating g_sync_event");
        }

        exit_status = xrdp_listen_main_loop(g_listen);
    }

    xrdp_listen_delete(g_listen);

    tc_mutex_delete(g_get_sync_mutex());
    g_set_sync_mutex(0);

    tc_mutex_delete(g_get_sync1_mutex());
    g_set_sync1_mutex(0);

    g_delete_wait_obj(g_get_term());
    g_set_term_event(0);

    g_delete_wait_obj(g_get_sigchld());
    g_set_sigchld_event(0);

    g_delete_wait_obj(g_get_sync_event());
    g_set_sync_event(0);

    if (daemon && (daemon_pid == g_getpid()))
    {
        /* Try to delete the PID file, although if we've dropped
         * privileges this won't be successful */
        g_file_delete(pid_file);
    }

    log_end();
    g_deinit();

    if (exit_status == 0)
    {
        g_exit(0);
    }
    else
    {
        g_exit(1);
    }

    return 0;
}
